k9mail/src/main/java/com/fsck/k9/fragment/MessageListFragment.java

```diff
@@ -2719,6 +2733,113 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
+                calendar.add(Calendar.DATE, -1);
+                if (calendar.get(calendar.DAY_OF_WEEK) == calendar.SUNDAY) {
+                    sundayOffset = j + 1;
+                    break;
+                }
+            }
+
+            Calendar messageDate = Calendar.getInstance();
+            messageDate.setTimeInMillis(date);
+
+            // calendar.setTime( compareDate );
+            long difference = daysBetween(messageDate, Calendar.getInstance());
+
+            // dateDifference( millisCompare, millisNow );
+            if (difference == 0) {
+                category = 1;
+            } else if (difference == 1) {
+                category = 2;
+            } else if (difference == 2) {
+                category = 3;
+            } else if ((sundayOffset > 1) && (difference <= sundayOffset)) {
+                category = messageDate.get(Calendar.DAY_OF_WEEK) + 3;
+            } else if ((difference - sundayOffset) < 7) {
+                category = 11;
+            } else if ((difference - sundayOffset) < 14) {
+                category = 12;
+            } else if ((difference - sundayOffset) < 21) {
+                category = 13;
+            } else {
+                category = 14;
+            }
+
+            return category;
+        }
+        return 0;
+    }
+
+    private final static long MILLISECS_PER_DAY = 24 * 60 * 60 * 1000;
+
+    private long daysBetween(Calendar a, Calendar b) {
+
+        // Optional: avoid cloning objects if it is the same day
+        if (a.get(Calendar.ERA) == b.get(Calendar.ERA) && a.get(Calendar.YEAR) == b.get(Calendar.YEAR)
+                && a.get(Calendar.DAY_OF_YEAR) == b.get(Calendar.DAY_OF_YEAR)) {
+            return 0;
+        }
+        Calendar a2 = (Calendar) a.clone();
+        Calendar b2 = (Calendar) b.clone();
+        a2.set(Calendar.HOUR_OF_DAY, 0);
+        a2.set(Calendar.MINUTE, 0);
+        a2.set(Calendar.SECOND, 0);
+        a2.set(Calendar.MILLISECOND, 0);
+        b2.set(Calendar.HOUR_OF_DAY, 0);
+        b2.set(Calendar.MINUTE, 0);
+        b2.set(Calendar.SECOND, 0);
+        b2.set(Calendar.MILLISECOND, 0);
+        long diff = a2.getTimeInMillis() - b2.getTimeInMillis();
+        long days = diff / MILLISECS_PER_DAY;
+        return Math.abs(days);
+    }
+
+    public int[] getMessageCategories() {
+        int category = -1;
+        int lastCategory = -1;
+
+        CursorAdapter hop = adapter;
+
+        int[] categories = new int[hop.getCount()];
+
+        for (int i = 0; i < hop.getCount(); i++) {
+            Cursor cursor = (Cursor) hop.getItem(i);
+
+            long date = 0;
+            if (sortType == SortType.SORT_DATE) {
+                date = cursor.getLong(DATE_COLUMN);
+            } else if (sortType == SortType.SORT_ARRIVAL) {
+                date = cursor.getLong(INTERNAL_DATE_COLUMN);
+            }
+
+            category = getDateCategory(date);
+            if (category != lastCategory) {
```

Comment:
Using an enum will make this `category == Category.NONE` and so much more understandable.